transport/src/main/java/io/netty/bootstrap/Bootstrap.java

```diff
@@ -184,44 +188,50 @@ public ChannelFuture connect(SocketAddress remoteAddress, SocketAddress localAdd
     /**
      * @see #connect()
      */
-    private ChannelFuture doResolveAndConnect(final SocketAddress remoteAddress, final SocketAddress localAddress) {
-        final ChannelFuture regFuture = initAndRegister();
-        final Channel channel = regFuture.channel();
+    private Future<Channel> doResolveAndConnect(final SocketAddress remoteAddress, final SocketAddress localAddress) {
+        EventLoop loop = group.next();
+        final Future<Channel> regFuture = initAndRegister(loop);
 
+        Promise<Channel> resolveAndConnectPromise = new DefaultPromise<>(loop);
         if (regFuture.isDone()) {
             if (!regFuture.isSuccess()) {
                 return regFuture;
             }
-            return doResolveAndConnect0(channel, remoteAddress, localAddress, channel.newPromise());
+            Channel channel = regFuture.getNow();
+            ChannelPromise promise = channel.newPromise();
```

Comment:
IIUC, this `ChannelPromise` is only required because we have to eventually call a pipeline method (`connect()` in this case) which requires ` ChannelPromise`. Can we do this adaptation `Promise<Channel>` to `ChannelPromise` as close to the call to the pipeline as possible so it is easy to understand the reason for doing so? In this case it will be `Bootstrap#doConnect()` I think.